File: android/rxflux-core/src/main/java/com/klefmusic/rxfluxcore/Reducer.kt
package com.klefmusic.rxfluxcore

import com.klefmusic.rxfluxcore.models.Action

/**
 * A function that returns the next state tree, given
 * the current state tree and the action to handle.
 */
typealias Reducer<S> = (oldState: S, action: Action) -> S

fun <S> reducer(wrapped: (state: S, action: Action) -> S): Reducer<S> = wrapped

fun <S> combineReducers(reducers: List<Reducer<S>>): Reducer<S> =
    reducer { state, action -> reducers.fold(state) { acc, r -> r(acc, action) } }

File: iOS/mobileflux/mobileflux/Reducer.swift
import Foundation
import RxSwift

public typealias Reducer<S: State> = (_ state: S, _ action: Action) -> S

func reducer<S: State>(_ reducer: @escaping Reducer<S> ) -> Reducer<S> {
    return reducer
}

func combineReducers<S: State>(_ reducers: Reducer<S>...)-> Reducer<S> {
    return reducer { state, action in
        reducers.reduce(state) { acc, r in r(acc, action) }
    }
}



